knowledge-base/changes-on-items-in-report-events-are-not-applied.md

```diff
@@ -0,0 +1,97 @@
+type: how-to
+page_title: Changes on items in report events are not applied
+slug: changes-on-items-in-report-events-are-not-applied
+res_type: kb
+---
+
+## Description
+
+Changes on items in report events are not applied.  
+
+In Telerik Reporting versions **prior to R3 2016**, all changes in events were accepted, even these applied on items' definitions. When items' definitions are modified in events, changes are applied on all consecutive processing elements with an expected result.  
+  
+Modifying report in events prior to **R3 2016:**  
+
+```cs
+void report_NeedDataSource(object sender, EventArgs e)
+{
+    this.DataSource = GetReportData();
+    this.table.DataSource = GetTableData();
+}
+```
+  
+**As of [R3 2016](../upgrade-path-2016-r3#api-breaking-changes)**, we applied an optimization change in the report rendering algorithm and now each definition item property values are read and cached once the report rendering starts.   
+
+If you are working with item definitions in events, they will **not** be modified. For more details, check the updated [Understanding Events](../understanding-events)help article.
+
+Modifying report in events **after R3 2016:**  
+
+```cs
+void report_NeedDataSource(object sender, EventArgs e)
+{
+    Telerik.Reporting.Processing.Report processingReport = (Telerik.Reporting.Processing.Report)sender;
+    processingReport.DataSource = GetReportData();
+}
+```
+
+  
+> **Note**
+> <br/>
+> In the report **NeedDataSource** and **ItemDataBinding** events, the processing instances of the other report items are not created yet. Hence, it is not possible to modify other items in these Report events.  
+ 
+Modifying other items in events **after R3 2016:**  
+
+```cs
+void table_NeedDataSource(object sender, EventArgs e)
+{
+    Telerik.Reporting.Processing.Table processingTable = (Telerik.Reporting.Processing.Table)sender;
+    processingTable.DataSource = GetTableData();
+}
+```
+
+## Solution
+
+In general, in all versions of Telerik Reporting, changes in events are possible if they are applied on the **processing elements**, not the items' definitions (report elements accessible through **this.** in C# or **Me.** in VB.NET). The valid approaches are illustrated in [Accessing Report Items Programmatically](../faq-accessing-items).  
+  
+The recommended approach is to use the *expression engine* and to avoid custom code in events:  
+
+1. [Conditional Formatting](../styling-conditional-formatting) to control styles via expressions;
+2. [Bindings](../expressions-bindings) to set properties of items via expressions;
+3. [User functions](../expressions-user-functions) to set properties of items, when the [built-in functions](../expressions-reference) are not sufficient.
+
+The usage of the expression engine assures items will be modified at the right moment.  
+  
+Interesting help articles related to binding data via expressions and modifying reports:  
```

Comment:
All the below links can go to a See Also section of the KB. This will keep the solution section clean and will also help navigating to other helpful resources.
https://wiki.progress.com/display/DTS/Knowledge+Base+Content+Standards#KnowledgeBaseContentStandards-How-ToArticles